compiler/rustc_macros/src/diagnostics/diagnostic.rs

```diff
@@ -151,3 +160,46 @@ impl<'a> LintDiagnosticDerive<'a> {
         })
     }
 }
+
+/// Checks whether the slug starts with the crate name it's in.
+fn make_check(slug: &syn::Path) -> TokenStream {
```

Comment:
This const string searching routine is pretty crude and the errors SUCK, so I'm happy to cut this out if you think it'd be better to not have this.

Reply:
This check basically enforces the change you're making in this PR, right?

I think that's a good thing to have. How hard is it to get an error here? Who sees the errors (e.g., is it just rustc devs)? I'm assuming there's nothing fundamental that makes improving the errors in the future hard? Does this have the potential to significantly hurt compile times for either rustc, or general Rust programs?

If this is something that only affects rustc devs and helps them maintain this pattern going forward, then I think enforcement is worth keeping even if it's crude.

Reply:
This check seems much more complicated than it needs to be.

I would have thought that we could do something like this to get the current crate name..

```rust
let crate_name = std::env::var("CARGO_PKG_NAME")
        // If `CARGO_PKG_NAME` is missing, then we're probably running in a test, so use
        // `no_crate`.
        .unwrap_or_else(|_| "no_crate".to_string())
        .replace("rustc_", "");
```
(from [#103042 (diff)](https://github.com/rust-lang/rust/pull/103042/files#diff-a329c61992d948b81c129613c0a816d2f0ebac479fc85bd92b7ab3b2139b7bf3R71-R75))

..and then just do `starts_with` comparison on the slug we get from the user? 

All of which would happen in the macro itself, rather than in the macro-generated code, and could emit nice diagnostics like the rest of the macro?

Reply:
@davidtwco I have implemented this change